api/src/display.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+//! Character encodings for public addresses and request types for easy
+//! cut-and-paste.
+
+use crate::printable;
+use crc::crc32;
+use failure::Fail;
+use protobuf::{parse_from_bytes, Message};
+
+/// Decoding / encoding errors
+#[derive(Clone, Debug, Eq, PartialEq, Fail)]
+pub enum Error {
+    /// Protocol buffer could not be serialized
+    #[fail(display = "Could not serialize protocol buffer: {:?}", _0)]
+    SerializationFailure(String),
+
+    /// The b58 string cannot be converted into bytes
+    #[fail(display = "Could not decode b58: {:?}", _0)]
+    B58Error(String),
+
+    /// Protocol buffer could not be deserialized
+    #[fail(display = "Could not deserialize protocol buffer: {:?}", _0)]
+    DeserializationFailure(String),
+
+    /// Checksum does not match
+    #[fail(display = "Checksum does not match payload")]
+    ChecksumFailure(),
+}
+
+/// A little-endian IEEE CRC32 checksum is prepended to payloads.
+fn calculate_checksum(data: &[u8]) -> [u8; 4] {
+    crc32::checksum_ieee(data).to_le_bytes()
+}
+
+/// The B58 wrapper supports encoding the protocol buffer bytes as a b58
+/// encoded string, with a checksum prepended to it.
+impl printable::PrintableWrapper {
+    /// Converts the proto to bytes and then encodes as b58
+    pub fn b58_encode(&self) -> Result<String, Error> {
+        let wrapper_bytes = self
+            .write_to_bytes()
+            .map_err(|err| Error::SerializationFailure(err.to_string()))?;
+        let mut bytes_vec = Vec::new();
+        bytes_vec.extend_from_slice(&calculate_checksum(&wrapper_bytes));
+        bytes_vec.extend_from_slice(&wrapper_bytes);
+        Ok(bs58::encode(&bytes_vec[..]).into_string())
+    }
+
+    /// Converts a b58 string to bytes and then decodes to a proto
+    pub fn b58_decode(encoded: String) -> Result<Self, Error> {
+        let mut decoded_bytes = bs58::decode(encoded)
+            .into_vec()
+            .map_err(|err| Error::B58Error(err.to_string()))?;
+        let wrapper_bytes = decoded_bytes.split_off(4);
+        let expected_checksum = calculate_checksum(&wrapper_bytes);
+        if expected_checksum.to_vec() != decoded_bytes {
+            return Err(Error::ChecksumFailure());
+        }
+        let wrapper = parse_from_bytes(&wrapper_bytes)
+            .map_err(|err| Error::DeserializationFailure(err.to_string()))?;
+        Ok(wrapper)
+    }
+}
+
+#[cfg(test)]
+mod display_tests {
+    use crate::{
+        external,
+        printable::{PaymentRequest, PrintableWrapper},
+    };
+
```

Comment:
Can we include a test here to demonstrate that the selected base58 library uses the desired alphabet?

```
    /// Make sure that our base58 encoding uses the exprected bitcoin base58 symbols:
    /// "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    #[test]
    fn base58_encodes_as_expected() {
        let pairs: Vec<(&str, &str)> = vec![
            ("", ""),
            ("00", "1"),
            ("000ff2cdff6c2148", "1c47pDn6ib"),
            (
                "0070d785ee0df1e3d16b6a66f2204ca9299b2f2f12b222ef1ba0fcc80d57ac4c045100",
                "13YxPSL4Rf6Ss7NaG4tsBHhZ8tdsgmJtoz8eY6uHFC8rRAHm",
            ),
            (
                "83532b4c994fae8a4b5776c1f17ed3c8335ae4de9d6a0000",
                "CyP9pioEYvLMji52APxa2TMHyCNTJ1pZM",
            ),
            (
                "007e942441ac7db664cb421c62a9fb860c1f866f0b1441ebbc9153d6adacabee148048195b20da",
                "1Kj2RBZwxMnWRtdWR3gxb7DkYafn6iiuCiLEJjkyXGdh8uiwYHzsK",
            ),
            (
                "00cd8d43e8fbab800d01711ed44ee2fbb9e216011dde0f67d67f2a05de8b0000",
                "148ntA8etPpqANGzNjHkdMSC6LoQp4voo4YhWkj9pHLT",
            ),
            (
                "000ffd793d34f0a4b6822e89a3cd2f3164d4e99c81ecb918f515bf2d9748dc00",
                "1F9DrjRyZRfnpe69vfcTbCBsDJUGFtBecsTwnSz6tGw",
            ),
            ("f46d5eb462b90970bc311c", "23cVANQRQmn2c9uD"),
            ("ec67ae84eefb00", "9xc2y7X1KV"),
            (
                "56221023a34880371c28b7bafa8bb33b53738f1511ba85ff36d1a1cd3e93",
                "JG1QZ95iNUzAJdSh35umCpYr528FMMvh2QB7v4oer",
            ),
            (
                "00009fb83298081e8d0be05c4588cdd7086d",
                "11LivfDpGrReCDj2ma9zoWiG",
            ),
            ("0000cb9f6c0000", "11PySFonP"),
            ("364ee56c3c35c0160000", "43y1bASdG3BNGP"),
            ("00d5", "14g"),
            (
                "1ae8a674e5e32483ec9c6d65ffcf3f487439e4dfd6727a8af80c5a5d21364a7ddf94217b",
                "CrMEn7gVVXaeVaWsPuNTasWq68Vp4TfyRJ1cNbbgCeb7AQmwg",
            ),
            ("002935985c", "1246Vfy"),
            (
                "00358a736ff4d5f2d34905afdd6fdfcc6f7b96134b3af511b0c0870000",
                "1YwVY9cVon6oqqaEqLxFAirwFiKaXC9dyS2bYaj",
            ),
            ("00002e3b00", "11GXeF"),
            (
                "e593f3c4bd1771ffa3b9c7dfa7012a5553433145e4108d9ffd16c85940271ab39467755a",
                "2k7G6gTxHZsG2tpy7qMxH8v64uKCZqUYrHWvYBX6ass7GzaWjj",
            ),
        ];
        for (hex, expected_b58) in pairs.iter() {
            let bytes = hex::decode(hex).unwrap();
            let decoded_bs58 = bs58::decode(expected_b58).into_vec().unwrap();
            assert_eq!(bytes, decoded_bs58);

            let b58 = bs58::encode(bytes).into_string();
            assert_eq!(b58, *expected_b58);
        }
    }
```